Alc controller patch disable

RehabMan opened the pull request:
This change allows disabling of AppleALC's patching for controllers.
We ran into a situation with unsupported HDAU 0x0a0c on the Lenovo Y50.
Details start here:
https://www.tonymacx86.com/threads/guide-lenovo-y50-uhd-or-1080p-using-clover-uefi.261723/page-2#post-1831046

Eventually, we determine that the patching for AppleHDAController to accommodate 0x0c0c (as 0x0a0c) the way AppleALC does it simply does not work as well as the FakePCIID_Intel_HDMI_Audio.kext technique.

By disabling that feature in AppleALC, we can use FakePCIID_Intel_HDMI_Audio.kext to do the 0x0c0c -> 0x0a0c mapping instead.

The way I decided to allow this is by detecting property 'alc-controller-patch-disable'=<01 00 00 00> (or any non-zero injected on HDAU). Having that property present will disable patching within AppleALC for that specific controller.  The way it is coded, it applies to any controller that might be supported/patched by AppleALC, not just HDAU.

I also did a bit of cleanup in AlcEnabler::grabControllers.

I also changed the tab settings for the project, such that tabs during editing are not dependent on Xcode system-wide preferences, but rather the project specifies it explicitly.

For the future, I might suggest that AppleALC spoof PCI IDs as is done in FakePCIID, but perhaps with the same technique used by WhateverGreen.  This would obviate the need for FakePCIID most of the time, as this example (0x0c0c->0x0a0c) would work properly.

RehabMan commented on AppleALC/kern_alc.cpp:
```diff
@@ -446,7 +449,11 @@ void AlcEnabler::processKext(KernelPatcher &patcher, size_t index, mach_vm_addre
-			
+
+			if (controllers[i]->nopatch) {
+				SYSLOG("alc", "skipping %lu controller %X:%X:%X due to alc-controller-patch-disable", i, controllers[i]->vendor, controllers[i]->device, controllers[i]->revision);
```
I like it for diagnostic purposes (for those following one of my guides), as a way to verify that everything is as I expect.

But now that it is working, and you have kind of verified that it is the "right way", I can make it debug only.

RehabMan commented on AppleALC.xcodeproj/project.pbxproj:
```diff
@@ -518,7 +519,7 @@
 				MODULE_START = "$(PRODUCT_NAME)_kern_start";
 				MODULE_STOP = "$(PRODUCT_NAME)_kern_stop";
-				MODULE_VERSION = 1.3.3;
+				MODULE_VERSION = 1.3.3d4;
```
I did this build as a d3 build, as it is the third 1.3.3 custom build that I've done and when looking at diagnostic files (kextstat), I need a way to distinguish between other builds of 1.3.3 that I have provided before.

For a final commit, I'll make it back to 1.3.3.

RehabMan commented on AppleALC/kern_alc.cpp:
```diff
@@ -54,10 +54,11 @@ void AlcEnabler::updateProperties() {
 				updateDeviceProperties(devInfo->audioBuiltinDigital, devInfo, "onboard-1", false);
-				uint32_t dev = 0, rev = 0;
+				uint32_t dev = 0, rev = 0, nopatch = 0;
+				WIOKit::getOSDataValue(devInfo->audioBuiltinDigital, "alc-controller-patch-disable", nopatch);
```
I was just going with existing convention of WhateverGreen, which for example, uses 32-bit integer for things like framebuffer-patch-enable, framebuffer-conX-enable, etc.

I would normally use presence, but it just wasn't the convention used by other kexts in the "Lilu family".

vit9696 replied:
I believe we were not too consistent here and there, as the kexts evolve and get more features. If you look here and there you will see that most of the entries go with getProperty though: `getProperty("enable-hdmi20")`, `getProperty("disable-metal")`, `getProperty("no-audio-fixconn")`. Best to change it and make a bool I think.

vit9696 commented:
>I'll look into implementing the WhateverGreen method of spoofing the PCI ids, which would make it similarly bullet proof as FakePCIID. But before I address this in AppleALC, I want to fix spoofing in WhateverGreen first :-)

Ouch, are there any real-world issues? I thought we made things pretty reliable with @lvs1974.

> No Changelog.md updates for 1.3.3 yet (even though there have been commits for it), which is why I didn't add any. I can start a log for 1.3.3 that has my changes it in at least.

Um, are you updated? I see quite a large changelog for 1.3.3 in master, and I hope we get it ready for a November release…

RehabMan commented:
> Ouch, are there any real-world issues? I thought we made things pretty reliable with @lvs1974.

Sorry for the off-topic, but I'll try to explain my plans here anyway...

The only spoof supported is CFL as KBL.
KBL as SKL is not supported (no patching will be done).
I'm not certain the way CFL as KBL is implemented is 100% either (was reviewing the code and saw some potential issues)

I plan to change the code such that all of the graphics kexts are "hooked", so it is not dependent on detected CPU.

This would allow:
- any kind of spoofing you wanted to try on any CPU
- (likely) dumping ig-platform data no matter the CPU you have
- no dependencies on CPU detection (because it is broken anyway... my i7-8559U is detected as KBL, not CFL)

And I think the end result will actually be simplified code...

> Um, are you updated? I see quite a large changelog for 1.3.3 in master, and I hope we get it ready for a November release…

I see that an update in just the last 5 days.  I've been pretty heads down, not in sync with commits only 5 days old.  I tend to avoid pulling when I'm working on things...  Will pull/merge before pushing my changes.  Thanks for the reminder.

vit9696 commented:
Oh, you meant generation to generation spoofing, not just spoofing `device-id` within a single generation, which was initially intended.

In general, I believe, spoofing between the generations does not work good enough, and KBL/CFL is basically the only option it was of any use, mainly due to hardware being the same. While technically beneficial, it is not very practical. After CFL drivers were released even KBL/CFL spoofing was planned for removal, but some complaints made us reconsider the choice with a single exception. In addition there was the VM case, where the reported CPU type had to be different from the actual CPU due to VM incompatibilities. This is what we have `lilucpu` boot argument for, and it pretty much means "this CPU is X, but the system does not know it" and intended for use on VMs.

The reason we check the CPU generation exists is to increase the performance of the code (we have to do disk i/o to perform patching) and simplify the logic. If you want to make the decision later (e.g. based on IGPU `device-id` or `ig-platform-id`), you will have to initially setup all the kexts for modification, and later disable them in `onPatcherLoad` step. Will not hurt the performance, but will complicate things a little I believe.

> dumping ig-platform data no matter the CPU you have

I believe `lilucpu` was meant to work around this issue, but I continually forget what was the problem last time it was tried. CC @headkaze.

> no dependencies on CPU detection (because it is broken anyway... my i7-8559U is detected as KBL, not CFL)

This looks like a mistake in Lilu, which I made due to no proper information in regards to CPUIDs here: https://en.wikichip.org/wiki/intel/microarchitectures/coffee_lake#CPUID. I tried finding CPUID info on your CPU but the stepping is little confusing: https://en.wikichip.org/wiki/intel/core_i7/i7-8559u, they say it is D0, and for Kaby they say J1 @_@. This value is only 4 bits, and it needs to be actually checked and added to alter the following line (regardless of new/old way of IGPU detection):
https://github.com/acidanthera/Lilu/blob/master/Lilu/Sources/kern_cpu.cpp#L108 Could you help me here?

RehabMan commented:
>  If you want to make the decision later (e.g. based on IGPU device-id or ig-platform-id), you will have to initially setup all the kexts for modification, and later disable them in onPatcherLoad step. Will not hurt the performance, but will complicate things a little I believe.

I will give it a try and see how it goes.  Would help with my understanding of the Lilu API in any case.

> Coffee Lake U should have 0x8E id with 0xA stepping, and Kaby Lake U has 0x8E id with 0x9 stepping. Could you add the check against 9 equality here just like below? This should fix your CPU detection.

Thanks, I will double check the id/stepping and if appropriate, try the change you suggest.
For now, I just used lilucpu=10 to work around the bug.

RehabMan commented:
> To reduce my load I gave you write access to acidanthera, commit once done. In general, please work on features (if any) in separate branches, but minor bugfixes and fuckups can be fixed directly to reduce the load.

Just to clarify.... does this mean you don't want the changes for this PR committed to master?  Or am I to commit directly to master for the next build.

Note: I have a separate branch locally; addressed all your issues above, (and renamed alc-controller-patch-disable to no-controller-patch to follow the no-audio-fixconn example), all commits squashed to a single.  It is ready to merge to master.

vit9696 commented:
Thx, commit both please. Directly to master is fine for minor changes. Lilu may need version bump if I remember correctly.

RehabMan commented:
Revisted commit pushed: 83379d2cd51d41ff5aa8986ad8eaefba08e31f91
Closing this PR.

headkaze commented:
> I believe `lilucpu` was meant to work around this issue, but I continually forget what was the problem last time it was tried. CC @headkaze.

lilucpu works for all Intel gen's we required to get dumps from. The rest are either inside their FB kext anyway or when I tried to dump from a broken kext (TGL). TGL kext has since been removed by Apple so at this point is a non-issue.
